File: algorithms/greedy/fractional-knapsack/fractional_knapsack.cpp
#include <iostream>
#include <vector>

float greedy_knapsack(std::vector<float>&,int,const std::vector<int>&,const std::vector<int>&);
void sort_data(std::vector<int>&,std::vector<int>&);

int main(void){
    std::cout << "Fractional Knapsack Algorithm using C++\n";

    std::vector<int> weight = {15,25,30,90,5,10,45,20,35,80};
    std::vector<int> value = {90,38,76,58,87,62,63,98,26,41};
    int n = 10;

    int k_cap = 200;
    std::vector<float> solution_vector(n,0.0);
    
    sort_data(weight,value);

    float total_profit = greedy_knapsack(solution_vector,k_cap,weight,value);

    std::cout << "\nFor the given knapsack with capacity" <<  k_cap  << "total profit is " << total_profit;

    std::cout << "\n\nSolution vector is: [ ";
    for(int i=0; i<n-1; i++){
        std::cout << solution_vector[i] << ", ";
    }
    std::cout << solution_vector[n-1] << " ]";
    
    return 0;
}

float greedy_knapsack(std::vector<float>& x, int m, const std::vector<int>& w, const std::vector<int>& p){
    float sum = 0.0;
    int rem_cap = m;
    int j=0;

    for(int i=0; i<w.size(); i++){
        x[j] = 0.0;
    }

    for(; j<w.size(); j++){
        if(w[j]>rem_cap){
            break;
        }
        x[j] = 1.0;
        rem_cap -= w[j];
        sum += (float)p[j];
    }

    if(rem_cap>0){
        x[j] = (float)rem_cap/w[j];
        sum += (float)p[j]*x[j];
    }

    return sum;
}

void sort_data(std::vector<int>& w, std::vector<int>& p){
    int n = w.size();

    float s1[n];
    int s2[n];
    int t[2][n];
    float temp1;
    int temp;

    for(int i=0; i<n; i++){
        s1[i] = (float) p[i]/w[i];
        s2[i] = i;
        t[0][i] = p[i];
        t[1][i] = w[i];
    }

    for(int i=0; i<n-1; i++){
        for(int j=0; j<n-i-1; j++){
            if(s1[j] < s1[j+1]){
                temp1 = s1[j];
                s1[j] = s1[j+1];
                s1[j+1] = temp1;

                temp = s2[j];
                s2[j] = s2[j+1];
                s2[j+1] = temp;
            }
        }
    }


    for(int i=0; i<n; i++){
        p[i] = t[0][s2[i]];
        w[i] = t[1][s2[i]];
    }

    return;
}

File: algorithms/greedy/fractional-knapsack/FractionalKnapsack.java
class FractionalKnapsack {


    public static void main(String[] args){
        System.out.println("Fractional Knapsack Algorithm using Java");

        int k_cap = 200;
        int weight[] = {15,25,30,90,5,10,45,20,35,80};
        int value[] = {90,38,76,58,87,62,63,98,26,41};
        int n = weight.length;

        float[] solution_vector = new float[n];

        sortData(weight, value);

        float total_profit = fractionalKnapsack(solution_vector,k_cap,weight,value);

        System.out.println("\nFor the given knapsack with capacity "+k_cap+", total profit is "+total_profit);

        System.out.print("\n\nSolution vector is: [ ");
        for(int i=0; i<n-1; i++){
            System.out.print(solution_vector[i] + ", ");
        }
        System.out.print(solution_vector[n-1] + " ]");

        return;
    }


    public static float fractionalKnapsack(float[] x, int m, int[] w, int[] p){
        int n = w.length;
        int rem_cap = m;
        int j;
        float sum = 0.0f;

        for(int i=0; i<n; i++){
            x[i] = 0.0f;
        }
        
        j=0;
        for(; j<n; j++){
            if(w[j]>rem_cap){
                break;
            }
            x[j] = 1.0f;
            rem_cap -= w[j];
            sum += p[j];
        }

        if(rem_cap > 0){
            x[j] = (float)rem_cap/w[j];
            sum += (float)p[j]*x[j];
        }

        return sum;
    }


    public static void sortData(int[] w, int[] p){
        int n = w.length;
        float s1[] = new float[n];
        int s2[] = new int[n];
        int t[][] = new int[2][n];
        float temp1;
        int temp;

        for(int i=0; i<n; i++){
            s1[i] = (float)p[i]/w[i];
            s2[i] = i;
            t[0][i] = p[i];
            t[1][i] = w[i];
        }

        for(int i=0; i<n-1; i++){
            for(int j=0; j<n-i-1; j++){
                if(s1[j] < s1[j+1]){
                    temp1 = s1[j];
                    s1[j] = s1[j+1];
                    s1[j+1] = temp1;

                    temp = s2[j];
                    s2[j] = s2[j+1];
                    s2[j+1] = temp;
                }
            }
        }

        for(int i=0; i<n; i++){
            p[i] = t[0][s2[i]];
            w[i] = t[1][s2[i]];
        }

        return;
    }


}